transport/httptransport/http_transport.go

```diff
@@ -217,20 +227,29 @@ func (t *transfer) execute(ctx context.Context) error {
 
 		// start the http transfer
 		remaining := t.dealInfo.DealSize - t.nBytesReceived
-		if err = t.doHttp(ctx, req, of, remaining); err == nil {
+		reqErr := t.doHttp(ctx, req, of, remaining)
+		if reqErr == nil {
 			// if there's no error, transfer was successful
 			break
 		}
 		_ = of.Close()
+
+		// check if the error is a 4xx error, meaning there is a problem with
+		// the request (eg 401 Unauthorized)
+		if reqErr.code/100 == 4 {
```

Comment:
If the server is down you won't be able to connect to it (connection error).
HTTP 4xx indicates that there is something wrong with the format of the client's request, and if the same request were repeated it would get the same response:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses